API change: Return Promise instead of function

acostalima opened the pull request:
Resolves #7

satazor commented:
As per #13 discussion, I think that a `unhook` method on the returned promise is missing. 

Is that right @sindresorhus ?

sindresorhus commented on index.js:
```diff
@@ -32,19 +33,23 @@ function hook(type, opts, cb) {
-	};
+	return new Promise((resolve, reject) => {
+		unhookStdout.then(type => resolve(type)).catch(() => reject());
+		unhookStderr.then(type => resolve(type)).catch(() => reject());
```
You can't use `resolve` twice.

sindresorhus commented on index.js:
```diff
@@ -32,19 +33,23 @@ function hook(type, opts, cb) {
+
+module.exports.stderr = (opts, cb) => {
+	return hook('stderr', opts, cb);
 };
```
These could be inline arrow functions.

```js
module.exports.stderr = (opts, cb) => hook('stderr', opts, cb);
```

sindresorhus commented:
> I think that a unhook method on the returned promise is missing. Is that right @sindresorhus ?

Correct

acostalima commented:
Thanks for the feedback @sindresorhus  @satazor 
The implementation should be as desired now. I have also reverted the existing tests to `test.serial` and `test.serial.cb` and added a few more specifically to verify that the promise resolves when `unhook` is called either from promise or callback.

sindresorhus commented on readme.md:
```diff
@@ -62,7 +76,7 @@ Default: `false`
+##### transform
 
 Type: `Function`
 
```
Needs to document the `unhook` parameter.

acostalima replied:
Yeah, I missed that. Will do.

sindresorhus commented on readme.md:
````diff
@@ -16,35 +16,49 @@ $ npm install --save hook-std
+Unhook can also be done via callback:
+
+```js
+const promise = hookStd.stdout((output, encoding, unhook) => {
````
Drop the `encoding` parameter (from here and the code). I don't see the value of it in this module. This module is meant to operate on strings. Instead, just convert it to a string, if it's not already, before passing it into the `transform` function. If the `transform` function returns a string, convert it to the existing encoding, otherwise just pass it through with the correct encoding.

acostalima replied:
Could you please elaborate further on your last sentence?
Also, do you mean to actually not pass the `encoding` parameter to the transform function anymore?

sindresorhus commented on readme.md:
```diff
@@ -16,35 +16,49 @@ $ npm install --save hook-std
 Hook stdout and stderr.
 
-Returns a function that unhooks stdout and stderr.
+Returns a `Promise` that unhooks stdout and stderr.
```
This is not entirely correct. It returns a Promise that resolves when it's unhooked, and with a `.unhook()` method.

acostalima replied:
Agreed. I'll fix that.

sindresorhus commented on test.js:
```diff
@@ -80,6 +80,8 @@ function loggingWrite(log, retVal) {
 }
 
 test.serial('passes through the return value of the underlying write call', t => {
```
Instead of using `.serial` on every test, you can import AVA with this instead:

```js
import {serial as test} from 'ava';
```

acostalima replied:
I'll change that. Thanks.

sindresorhus commented on test.js:
```diff
@@ -213,3 +229,49 @@ test.serial('if once option is true and silent is false, hook only prints the fi
+	});
+	process.stderr.write('foo');
+	promise.unhook();
+	await promise.then(type => t.is(type, 'stderr'));
```
I don't really see the point in returning the `type`. I think it should just resolve with nothing.

Also to be able to, for example, return the collected output here.

acostalima replied:
Makes sense. I'll remove the return value.

sindresorhus commented on test.js:
```diff
@@ -213,3 +229,49 @@ test.serial('if once option is true and silent is false, hook only prints the fi
+test('promise resolves when stdout & stderr are hooked and released via promise', async t => {
+	t.plan(3);
+	const promise = m(str => {
+		if (str === 'foo' || str === 'bar') {
```
This test is too loose. I realize it would be caught by `t.plan` if one of the stream were not written to, but would be better to gather the output in an array for each call here and do a deepEqual assertion.

acostalima replied:
Ok.

acostalima commented:
Hey @sindresorhus, I completely forgot to finish this PR! My apologies. 
I made some improvements to the code since my last commit.
Please have look to the most recent changes I've made.

Thanks!

acostalima commented:
ping @sindresorhus

acostalima commented:
ping @sindresorhus

acostalima commented:
ping @sindresorhus

acostalima commented:
ping @sindresorhus

sindresorhus commented:
This looks great. Sorry for the long delay. I have more than 300 PRs to review.

![leo-celebrate](https://user-images.githubusercontent.com/170270/41173246-5c49a85e-6b80-11e8-87e1-5936814e915c.gif)
